### ISSUE TITLE
Dictionary-valued inputs/outputs/targets aren't correctly handled

### ISSUE THREAD

**Issue by ml-muse** [fastai/fastai#2088]
**Describe the bug**
Several functions fail to handle dictionary-type inputs, outputs, and/or targets. This includes the following:
- In torch_core.py, to_half() and to_float() don't recurse on dictionaries
- In basic_train.py, validate() fails to get the shape when yb is a dictionary
- In callback.py, AverageMetric.on_batch_end() fails to get the shape when yb is a dictionary
There may be more places with similar issues, but fixing these three was sufficient to allow basic training with a dictionary output.

**Provide your installation details**
fastai version 1.0.51

**To Reproduce**
```
import torch
from torch.utils.data import Dataset
import torch.nn as nn

import fastai
from fastai.train import Learner
from fastai.basic_data import DataBunch

class ExampleDataset(Dataset):
    def __init__(self):
        super(ExampleDataset, self).__init__()
        self.in1 = torch.randn(5, 4)
        self.in2 = torch.randn(5, 4)
        self.out1 = torch.randn(5, 3)
        self.out2 = torch.randn(5, 1)

    def __len__(self):
        return self.in1.shape[0]
    
    def __getitem__(self, index):
        return {'in1': self.in1[index], 'in2': self.in2[index]}, {'out1': self.out1[index], 'out2': self.out2[index]}

ds = ExampleDataset()
databunch = DataBunch.create(ds, ds, bs=1)

class ExampleModel(nn.Module):
    def __init__(self):
        super(ExampleModel, self).__init__()
        self.model1 = nn.Linear(4, 3)
        self.model2 = nn.Linear(4, 1)
        
    def forward(self, x):
        x = x['in1'] + x['in2']
        y1 = self.model1(x)
        y2 = self.model2(x)
        return {'out1': y1, 'out2': y2}

class ExampleLoss(nn.Module):
    def __init__(self):
        super(ExampleLoss, self).__init__()
        
    def forward(self, x_dict, y_dict):
        return nn.functional.mse_loss(x_dict['out1'], y_dict['out1']) + nn.functional.mse_loss(x_dict['out2'], y_dict['out2'])

learner = Learner(databunch, ExampleModel(), loss_func=ExampleLoss(), metrics=[]).to_fp16()

learner.fit_one_cycle(1)
```

**Expected behavior**
Model successfully trains for one epoch with no errors.

**Comment by sgugger** [comment]
First one can be fixed easily, but the next two are way trickier since the fastai library needs to get the batch size somehow. Note that the library doesn't handle listy outputs either (unless you use a callback to reduce that to a single output).

Are all your `ys` (or at least the first) with the batch size as the first dimension? Maybe we could have something that recursively tries to get the first element then takes it's first dim.

**Comment by ml-muse** [comment]
> First one can be fixed easily, but the next two are way trickier since the fastai library needs to get the batch size somehow. Note that the library doesn't handle listy outputs either (unless you use a callback to reduce that to a single output).

I was able to get this working locally with dictionaries by adding a bit of hacky code to take the batch size from yb.values()[0]. Interestingly I think the issue with listy outputs is different, as fastai does get the batch size from the first element, but instead fails due to uses of *xb in functions attempting to unpack the list into arguments. Though I haven't tried patching those, so it might still hit an issue getting the batch size later.

> Are all your ys (or at least the first) with the batch size as the first dimension? Maybe we could have something that recursively tries to get the first element then takes it's first dim.

Yes, and it seems reasonable to require the batch size be the first dimension of the ys. So that recursive function to compute batch_size sounds great.

Thanks!

**Comment by sgugger** [comment]
I think it should be fixed now. Let me know if you encounter any problem!
